### ISSUE TITLE
Active Storage saves a blob even when the model doesn't save due to failed validation

### ISSUE THREAD

**Issue by ConfusedVorlon** [rails/rails#36080]
### Steps to reproduce
```
class User < ActiveRecord::Base
  has_one_attached :avatar
  validates :name, presence: true, allow_blank: false

end

@user = User.create(avatar: AN_IMAGE)

```

### Expected behavior

@user doesn't save as it fails validation
no additional attachments or blobs added to database

### Actual behavior

@user doesn't save as it fails validation
new blob is added to database

### System configuration
**Rails version**:
rails (5.2.3)

**Ruby version**:
ruby-2.4.0

**Comment by ConfusedVorlon** [comment]
note - I have tried to workaround this with a callback

```
after_validation :delete_blob_if_validation_failed

  def delete_blob_if_validation_failed
    if self.errors.any? && !self.persisted?
      avatar&.purge
    end
  end
```

This deletes the backing file from storage, but doesn't delete the record from the active_storage_blobs table

**Comment by georgeclaghorn** [comment]
This is fixed in master and the 6.0.0 betas.

**Comment by ConfusedVorlon** [comment]
thanks

**Comment by mopx** [comment]
Is this a Rails 5.2.3 bug? can I rollback to Rails 5.2.2?

**Comment by gerryhd** [comment]
Is there no 5.2.x version that supports it?

**Comment by KapilSachdev** [comment]
The reported issue fix should had been back-ported in `5.2` as this is a wrong behaviour of rails.
But as `5.2` does not receives any more bug fixes as mentioned in [maintenance_policy](https://guides.rubyonrails.org/maintenance_policy.html#bug-fixes), I don't think rails team will be interested in releasing this in `5.2` version.

You should use `valid?` and then `save` calls instead of create but this might still be an issue if rollback occurs by any means after validation, or may be monkey patch this.
